<a href="https://colab.research.google.com/github/Srijan-Rai/Zomato-Restaurant-Clustering-Sentiment-Analysis/blob/main/ZOMATO_RESTAURANT_CLUSTERING_AND_SENTIMENT_ANALYSIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement**

Zomato is an Indian restaurant aggregator and food delivery start-up founded by Deepinder Goyal and Pankaj Chaddah in 2008. Zomato provides information, menus and user-reviews of restaurants, and also has food delivery options from partner restaurants in select cities.

India is quite famous for its diverse multi cuisine available in a large number of restaurants and hotel resorts, which is reminiscent of unity in diversity. Restaurant business in India is always evolving. More Indians are warming up to the idea of eating restaurant food whether by dining outside or getting food delivered. The growing number of restaurants in every state of India has been a motivation to inspect the data to get some insights, interesting facts and figures about the Indian food industry in each city. So, this project focuses on analysing the Zomato restaurant data for each city in India.

The Project focuses on Customers and Company, you have  to analyze the sentiments of the reviews given by the customer in the data and made some useful conclusion in the form of Visualizations. Also, cluster the zomato restaurants into different segments. The data is vizualized as it becomes easy to analyse data at instant. The Analysis also solve some of the business cases that can directly help the customers finding the Best restaurant in their locality and for the company to grow up and work on the fields they are currently lagging in.

This could help in clustering the restaurants into segments. Also the data has valuable information around cuisine and costing which can be used in cost vs. benefit analysis

Data could be used for sentiment analysis. Also the metadata of reviewers can be used for identifying the critics in the industry. 

# **Attribute Information**

## **Zomato Restaurant names and Metadata**
Use this dataset for clustering part

1. Name : Name of Restaurants

2. Links : URL Links of Restaurants

3. Cost : Per person estimated Cost of dining

4. Collection : Tagging of Restaurants w.r.t. Zomato categories

5. Cuisines : Cuisines served by Restaurants

6. Timings : Restaurant Timings

## **Zomato Restaurant reviews**
Merge this dataset with Names and Matadata and then use for sentiment analysis part

1. Restaurant : Name of the Restaurant

2. Reviewer : Name of the Reviewer

3. Review : Review Text

4. Rating : Rating Provided by Reviewer

5. MetaData : Reviewer Metadata - No. of Reviews and followers

6. Time: Date and Time of Review

7. Pictures : No. of pictures posted with review

# ZOMATO RESTAURANT CLUSTERING AND SENTIMENT ANALYSIS

In [ ]:
# Importing initial necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Importing drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Data paths
metadata_path = "/content/drive/MyDrive/Colab Notebooks/Zomato Clustering_Sentiment Analysis/Zomato Restaurant names and Metadata.csv"
reviews_path = "/content/drive/MyDrive/Colab Notebooks/Zomato Clustering_Sentiment Analysis/Zomato Restaurant reviews.csv"

In [ ]:
# Creating dataframes
metadata = pd.read_csv(metadata_path)
reviews = pd.read_csv(reviews_path)

In [ ]:
# metadata columns
metadata.columns

Index(['Name', 'Links', 'Cost', 'Collections', 'Cuisines', 'Timings'], dtype='object')

In [ ]:
# reviews coulmns
reviews.columns

Index(['Restaurant', 'Reviewer', 'Review', 'Rating', 'Metadata', 'Time',
       'Pictures'],
      dtype='object')

## Head

In [ ]:
metadata.head()

Name  \
0                 Beyond Flavours   
1                        Paradise   
2                        Flechazo   
3  Shah Ghouse Hotel & Restaurant   
4      Over The Moon Brew Company   

                                               Links   Cost  \
0  https://www.zomato.com/hyderabad/beyond-flavou...    800   
1  https://www.zomato.com/hyderabad/paradise-gach...    800   
2  https://www.zomato.com/hyderabad/flechazo-gach...  1,300   
3  https://www.zomato.com/hyderabad/shah-ghouse-h...    800   
4  https://www.zomato.com/hyderabad/over-the-moon...  1,200   

                                         Collections  \
0  Food Hygiene Rated Restaurants in Hyderabad, C...   
1                                Hyderabad's Hottest   
2                 Great Buffets, Hyderabad's Hottest   
3                             Late Night Restaurants   
4  Best Bars & Pubs, Food Hygiene Rated Restauran...   

                                            Cuisines  \
0  Chinese, Continental, Kebab, European, South I...   
1                     Biryani, North Indian, Chinese   
2       Asian, Mediterranean, North Indian, Desserts   
3  Biryani, North Indian, Chinese, Seafood, Bever...   
4  Asian, Continental, North Indian, Chinese, Med...   

                                             Timings  
0      12noon to 3:30pm, 6:30pm to 11:30pm (Mon-Sun)  
1                                     11 AM to 11 PM  
2              11:30 AM to 4:30 PM, 6:30 PM to 11 PM  
3                                    12 Noon to 2 AM  
4  12noon to 11pm (Mon, Tue, Wed, Thu, Sun), 12no...

In [ ]:
reviews.head()

Restaurant              Reviewer  \
0  Beyond Flavours     Rusha Chakraborty   
1  Beyond Flavours  Anusha Tirumalaneedi   
2  Beyond Flavours       Ashok Shekhawat   
3  Beyond Flavours        Swapnil Sarkar   
4  Beyond Flavours                Dileep   

                                              Review Rating  \
0  The ambience was good, food was quite good . h...      5   
1  Ambience is too good for a pleasant evening. S...      5   
2  A must try.. great food great ambience. Thnx f...      5   
3  Soumen das and Arun was a great guy. Only beca...      5   
4  Food is good.we ordered Kodi drumsticks and ba...      5   

                  Metadata             Time  Pictures  
0   1 Review , 2 Followers  5/25/2019 15:54         0  
1  3 Reviews , 2 Followers  5/25/2019 14:20         0  
2  2 Reviews , 3 Followers  5/24/2019 22:54         0  
3    1 Review , 1 Follower  5/24/2019 22:11         0  
4  3 Reviews , 2 Followers  5/24/2019 21:37         0

## Tail

In [ ]:
metadata.tail()

Name  \
100                               IndiBlaze   
101                            Sweet Basket   
102                        Angaara Counts 3   
103                             Wich Please   
104  Republic Of Noodles - Lemon Tree Hotel   

                                                 Links   Cost  \
100  https://www.zomato.com/hyderabad/indiblaze-gac...    600   
101  https://www.zomato.com/hyderabad/sweet-basket-...    200   
102  https://www.zomato.com/hyderabad/angaara-count...    500   
103  https://www.zomato.com/hyderabad/wich-please-1...    250   
104  https://www.zomato.com/hyderabad/republic-of-n...  1,700   

              Collections                         Cuisines  \
100                   NaN                 Fast Food, Salad   
101                   NaN                   Bakery, Mithai   
102                   NaN   North Indian, Biryani, Chinese   
103                   NaN                        Fast Food   
104  Pan-Asian Delicacies  Thai, Asian, Chinese, Malaysian   

                                               Timings  
100                                     11 AM to 11 PM  
101  10 AM to 10 PM (Mon-Thu), 8 AM to 10:30 PM (Fr...  
102                                   12 Noon to 11 PM  
103                           8am to 12:30AM (Mon-Sun)  
104              11:30 AM to 3 PM, 7 PM to 12 Midnight

In [ ]:
reviews.tail()

Restaurant          Reviewer  \
9995  Chinese Pavilion  Abhishek Mahajan   
9996  Chinese Pavilion    Sharad Agrawal   
9997  Chinese Pavilion         Ramandeep   
9998  Chinese Pavilion   Nayana Shanbhag   
9999  Chinese Pavilion  Bhargava Krishna   

                                                 Review Rating  \
9995  Madhumathi Mahajan Well to start with nice cou...      3   
9996  This place has never disappointed us.. The foo...    4.5   
9997  Bad rating is mainly because of "Chicken Bone ...    1.5   
9998  I personally love and prefer Chinese Food. Had...      4   
9999  Checked in here to try some delicious chinese ...    3.5   

                          Metadata             Time  Pictures  
9995     53 Reviews , 54 Followers    6/5/2016 0:08         0  
9996      2 Reviews , 53 Followers   6/4/2016 22:01         0  
9997    65 Reviews , 423 Followers   6/3/2016 10:37         3  
9998    13 Reviews , 144 Followers  5/31/2016 17:22         0  
9999  472 Reviews , 1302 Followers  5/31/2016 16:41         6

## Data Analysis

Let us carry out some data preprocessing and data visualization on the given data, to obtain clean data and also to gain some more insights into the data before we start with the main objectives of our project.

### Info

In [ ]:
metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         105 non-null    object
 1   Links        105 non-null    object
 2   Cost         105 non-null    object
 3   Collections  51 non-null     object
 4   Cuisines     105 non-null    object
 5   Timings      104 non-null    object
dtypes: object(6)
memory usage: 5.0+ KB


* In the metadata dataframe, there are no numerical features.
* The shape of this dataframe is (105,6). Hence there are 105 entries and 6 features.
* As we could see, some of the features contain null values. These null values need to be handled further

In [ ]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Restaurant  10000 non-null  object
 1   Reviewer    9962 non-null   object
 2   Review      9955 non-null   object
 3   Rating      9962 non-null   object
 4   Metadata    9962 non-null   object
 5   Time        9962 non-null   object
 6   Pictures    10000 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 547.0+ KB


* In the reviews dataframe, only Pictures is the feature with the numerical values
* The shape of this dataframe is (10000, 7), hence there are 10000 entries and 7 features
* There are some features that contain null values.

### Description

Let us get some basic description of the dataframes

In [ ]:
metadata.describe(include ="all")

Name                                              Links  \
count               105                                                105   
unique              105                                                105   
top     Beyond Flavours  https://www.zomato.com/hyderabad/beyond-flavou...   
freq                  1                                                  1   

       Cost                                  Collections  \
count   105                                           51   
unique   29                                           42   
top     500  Food Hygiene Rated Restaurants in Hyderabad   
freq     13                                            4   

                     Cuisines         Timings  
count                     105             104  
unique                     92              77  
top     North Indian, Chinese  11 AM to 11 PM  
freq                        4               6

In [ ]:
reviews.describe(include ="all")

Restaurant     Reviewer Review Rating  Metadata             Time  \
count             10000         9962   9955   9962      9962             9962   
unique              100         7446   9364     10      2477             9782   
top     Beyond Flavours  Parijat Ray   good      5  1 Review  7/29/2018 20:34   
freq                100           13    237   3832       919                3   
mean                NaN          NaN    NaN    NaN       NaN              NaN   
std                 NaN          NaN    NaN    NaN       NaN              NaN   
min                 NaN          NaN    NaN    NaN       NaN              NaN   
25%                 NaN          NaN    NaN    NaN       NaN              NaN   
50%                 NaN          NaN    NaN    NaN       NaN              NaN   
75%                 NaN          NaN    NaN    NaN       NaN              NaN   
max                 NaN          NaN    NaN    NaN       NaN              NaN   

            Pictures  
count   10000.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean        0.748600  
std         2.570381  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max        64.000000

### Duplicates in the data

In [ ]:
metadata.duplicated().sum()

0

There are no duplicate entries in the metadata dataframe.

In [ ]:
reviews.duplicated().sum()

36

Since there are some duplicates in the dataframe, we would further review these entries and decide the necessary action that needs to be taken on these duplicates.

In [ ]:
reviews[reviews.duplicated()]

Restaurant Reviewer Review Rating Metadata Time  Pictures
8778  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8779  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8780  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8781  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8782  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8783  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8784  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8785  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8786  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8787  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8788  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8789  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8790  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8791  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8792  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8793  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8794  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8795  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8796  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8797  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8798  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
8799  American Wild Wings      NaN    NaN    NaN      NaN  NaN         0
9086         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9087         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9088         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9089         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9090         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9091         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9092         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9093         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9094         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9095         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9096         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9097         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9098         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0
9099         Arena Eleven      NaN    NaN    NaN      NaN  NaN         0

Since all the entries for the duplicate restaurants are null, we would be dropping these entries

In [ ]:
reviews.drop_duplicates(inplace = True)

### Null Values in the data

In [ ]:
metadata.isna().sum()

Name            0
Links           0
Cost            0
Collections    54
Cuisines        0
Timings         1
dtype: int64

As we could see there are many null values under the collections feature. These collections are different tags provided by zomato for different restaurants. With the help of these collections, it becomes easier to filter out the restaurants which belong to a particular restaurant collection.

For the null values under collections, we would replace them with "No Collection", because these restaurants belong to no collections.

In [ ]:
metadata["Collections"].fillna("No Collection", inplace = True)

Since there are only 2 null values under timings feature, we would drop these entries from the dataframe.

In [ ]:
metadata.dropna(subset = ["Timings"],inplace = True)

In [ ]:
reviews.isna().sum()

Restaurant    0
Reviewer      2
Review        9
Rating        2
Metadata      2
Time          2
Pictures      0
dtype: int64

Let us drop all the null values which are present in the reviews data frame, as there are very less number of null values in the data frame.

In [ ]:
reviews.dropna(inplace= True)

### Feature Engineering

Let us perform some feature engineering on the available datasets, which will help us to visualize the data easily and more efficiently. This step includes cleaning the data under different features, adding necessary features into the dataset by using the existing features, converting the datatypes of the features, etc.

#### Feature Engineering on Metadata

In [ ]:
# Metadata cost datatype conversion
metadata["Cost"] = metadata["Cost"].str.replace(",","")
metadata["Cost"] = metadata["Cost"].astype("float64")

From the above step we converted the data type of "Cost

In [ ]:
def num_of_Cusines(cuisines):
  """ Function to obtain the number of cuisines
      Parameter: Cuisines under the metadata dataframe"""
  cuisinelist = list(str(cuisines).split(","))
  return len(cuisinelist)

In [ ]:
metadata["Number_of_Cusines"] = metadata["Cuisines"].apply(num_of_Cusines)

The above step creates a new feature called the number of cuisines, to obtain the number of cuisines served in a restaurant.

In [ ]:
metadata.drop(["Links","Timings"], axis = 1, inplace = True)
metadata.rename(columns={"Name":"Restaurant"},inplace = True)

The above step drops the Links and Timings features which are not necessary for our further analysis.

The Name feature is renamed as Restaurant.

#### Feature Engineering on reviews data

In [ ]:
reviews = reviews[reviews["Rating"] != "Like"]
reviews["Rating"] = reviews['Rating'].astype("float64")

Te above step converts the datatype of Rating Feature to float.

In [ ]:
reviews["Time"] = pd.to_datetime(reviews["Time"])
reviews["Day"] = reviews["Time"].dt.day_name()
reviews["Hour"] = reviews["Time"].dt.hour

Converting the datatype of "Time" to datetime and creating new features to get the day and hour from it.

In [ ]:
# Splitting the metadata into two new features
reviews['No_of_Reviews'],reviews['No_of_Followers']=reviews['Metadata'].str.split(',').str
reviews['No_of_Reviews'] = pd.to_numeric(reviews['No_of_Reviews'].str.split(' ').str[0])
reviews['No_of_Followers'] = pd.to_numeric(reviews['No_of_Followers'].str.split(' ').str[1])

Obtaining the number of reviews and followers of the restaurant reviews.

#### Datasets after feature Engineering

In [ ]:
metadata

Restaurant    Cost  \
0                           Beyond Flavours   800.0   
1                                  Paradise   800.0   
2                                  Flechazo  1300.0   
3            Shah Ghouse Hotel & Restaurant   800.0   
4                Over The Moon Brew Company  1200.0   
..                                      ...     ...   
100                               IndiBlaze   600.0   
101                            Sweet Basket   200.0   
102                        Angaara Counts 3   500.0   
103                             Wich Please   250.0   
104  Republic Of Noodles - Lemon Tree Hotel  1700.0   

                                           Collections  \
0    Food Hygiene Rated Restaurants in Hyderabad, C...   
1                                  Hyderabad's Hottest   
2                   Great Buffets, Hyderabad's Hottest   
3                               Late Night Restaurants   
4    Best Bars & Pubs, Food Hygiene Rated Restauran...   
..                                                 ...   
100                                      No Collection   
101                                      No Collection   
102                                      No Collection   
103                                      No Collection   
104                               Pan-Asian Delicacies   

                                              Cuisines  Number_of_Cusines  
0    Chinese, Continental, Kebab, European, South I...                  6  
1                       Biryani, North Indian, Chinese                  3  
2         Asian, Mediterranean, North Indian, Desserts                  4  
3    Biryani, North Indian, Chinese, Seafood, Bever...                  5  
4    Asian, Continental, North Indian, Chinese, Med...                  5  
..                                                 ...                ...  
100                                   Fast Food, Salad                  2  
101                                     Bakery, Mithai                  2  
102                     North Indian, Biryani, Chinese                  3  
103                                          Fast Food                  1  
104                    Thai, Asian, Chinese, Malaysian                  4  

[104 rows x 5 columns]

In [ ]:
reviews

Restaurant              Reviewer  \
0      Beyond Flavours     Rusha Chakraborty   
1      Beyond Flavours  Anusha Tirumalaneedi   
2      Beyond Flavours       Ashok Shekhawat   
3      Beyond Flavours        Swapnil Sarkar   
4      Beyond Flavours                Dileep   
...                ...                   ...   
9995  Chinese Pavilion      Abhishek Mahajan   
9996  Chinese Pavilion        Sharad Agrawal   
9997  Chinese Pavilion             Ramandeep   
9998  Chinese Pavilion       Nayana Shanbhag   
9999  Chinese Pavilion      Bhargava Krishna   

                                                 Review  Rating  \
0     The ambience was good, food was quite good . h...     5.0   
1     Ambience is too good for a pleasant evening. S...     5.0   
2     A must try.. great food great ambience. Thnx f...     5.0   
3     Soumen das and Arun was a great guy. Only beca...     5.0   
4     Food is good.we ordered Kodi drumsticks and ba...     5.0   
...                                                 ...     ...   
9995  Madhumathi Mahajan Well to start with nice cou...     3.0   
9996  This place has never disappointed us.. The foo...     4.5   
9997  Bad rating is mainly because of "Chicken Bone ...     1.5   
9998  I personally love and prefer Chinese Food. Had...     4.0   
9999  Checked in here to try some delicious chinese ...     3.5   

                          Metadata                Time  Pictures       Day  \
0           1 Review , 2 Followers 2019-05-25 15:54:00         0  Saturday   
1          3 Reviews , 2 Followers 2019-05-25 14:20:00         0  Saturday   
2          2 Reviews , 3 Followers 2019-05-24 22:54:00         0    Friday   
3            1 Review , 1 Follower 2019-05-24 22:11:00         0    Friday   
4          3 Reviews , 2 Followers 2019-05-24 21:37:00         0    Friday   
...                            ...                 ...       ...       ...   
9995     53 Reviews , 54 Followers 2016-06-05 00:08:00         0    Sunday   
9996      2 Reviews , 53 Followers 2016-06-04 22:01:00         0  Saturday   
9997    65 Reviews , 423 Followers 2016-06-03 10:37:00         3    Friday   
9998    13 Reviews , 144 Followers 2016-05-31 17:22:00         0   Tuesday   
9999  472 Reviews , 1302 Followers 2016-05-31 16:41:00         6   Tuesday   

      Hour  No_of_Reviews  No_of_Followers  
0       15              1              2.0  
1       14              3              2.0  
2       22              2              3.0  
3       22              1              1.0  
4       21              3              2.0  
...    ...            ...              ...  
9995     0             53             54.0  
9996    22              2             53.0  
9997    10             65            423.0  
9998    17             13            144.0  
9999    16            472           1302.0  

[9954 rows x 11 columns]